eth_utils/hexidecimal.py

```diff
@@ -2,36 +2,41 @@
 
 import binascii
 import codecs
-import string
 
 from .types import (
-    is_string,
     is_bytes,
+    is_string,
 )
-from .string import (
-    coerce_args_to_bytes,
-    coerce_return_to_text,
-    coerce_return_to_bytes,
-    force_obj_to_text,
+from .conversions import (
+    to_bytes,
+    to_text,
 )
 from .formatting import (
     is_prefixed,
 )
 
 
-@coerce_return_to_bytes
 def decode_hex(value):
+    """
+    Returns `value` decoded into a byte string.
+    Accepts any string with or without the `0x` prefix.
+    """
     if not is_string(value):
         raise TypeError('Value must be an instance of str or unicode')
-    return codecs.decode(remove_0x_prefix(value), 'hex')
+    decoded = codecs.decode(remove_0x_prefix(value), 'hex')
+    return to_bytes(primitive=decoded)
 
 
-@coerce_args_to_bytes
-@coerce_return_to_text
-def encode_hex(value):
+def encode_hex(primitive=None, hexstr=None, text=None):
+    """
+    Returns any supported value encoded into a hexidecimal representation
+    with a `0x` prefix.
+    """
+    value = to_bytes(primitive, hexstr, text)
     if not is_string(value):
         raise TypeError('Value must be an instance of str or unicode')
-    return add_0x_prefix(codecs.encode(value, 'hex'))
+    encoded = add_0x_prefix(codecs.encode(value, 'hex'))
```

Comment:
Since we *know* that the return value of this only has characters in the ASCII range **and** that it's going to be a bytes type, we can just do this:

```python
return add_0x_prefix(codecs.encode(value, 'hex')).decode('ascii')
```

Which will return a text type.  @carver any reason that we should instead use `uft8`?

Reply:
I think this also removes your circular import problem.

Reply:
Almost, this method still uses `to_bytes` from `conversions.py`

Reply:
I don't see any reason it needs to be utf8. It's surprising to me that `add_0x_prefix` is returning a `bytes` value though.

Reply:
I think `add_0x_prefix` returns bytes or text...
`def add_0x_prefix(value):
    if is_0x_prefixed(value):
        return value

    prefix = b'0x' if is_bytes(value) else '0x'
    return prefix + value`

Reply:
I mean, divorced from any historical support, I find that surprising. I think that `codecs` made a very odd choice here of encoding bytes to a hex string, then encoding it using ascii/utf8 to bytes again. The much newer `bytes.hex()` seems to do the more sane thing here.

Maybe we can just drop the support for bytes in `add_0x_prefix`, and use this here:

```py
return add_0x_prefix(value.hex())
```

Reply:
+1 to dropping bytes support to `add_0x_prefix`.  I'm inclined to adopt the policy that all hex things *must* be text types.